src/cirrus/lib/task.py

```diff
@@ -0,0 +1,208 @@
+from abc import ABC, abstractmethod #abstractclassmethod
+import argparse
+from copy import deepcopy
+import json
+import logging
+from os import makedirs
+import os.path as op
+from shutil import rmtree
+import sys
+from tempfile import mkdtemp
+from typing import Dict, List, Optional
+
+from boto3utils import s3
+
+from cirrus.lib.logging import get_task_logger
+from cirrus.lib.process_payload import ProcessPayload
+from cirrus.lib.transfer import download_item_assets, upload_item_assets
+
+
+class Task(ABC):
+
+    _name = 'task'
+    _description = 'A task for doing things'
+    _version = '0.1.0'
+
+    def __init__(self: "Task", payload: Dict, local: Optional[bool]=None, workdir: Optional[str]=None):
+        # parse event
+        payload = ProcessPayload.from_event(payload)
+
+        # TODO - use PySTAC here
+        self.original_items = payload.pop('features')
+        self.items = deepcopy(self.original_items)
+
+        self.process_definition = payload.pop('process')
```

Comment:
For compatibility with workflow chaining, I think you want to use `payload.process` here, as that will figure out the current process definition in the payload for you. Otherwise you could get a list of process definitions.

Reply:
Well, I say that, but it seems this attribute is used two different ways. In light of that realization, I think maybe you want something like this:

```suggestion
        self.process_definition = payload.process
        self._process_definition = payload.pop('process')
```

Then in `output_payload` you would use `self._process_definition.

That said, I'd be interested in learning more about the need here and what problem(s) this whole pull-it-apart-and-rebuild-later strategy solves for us.

Reply:
Yeah, good points here.

The motivation here was to try and keep things clear for developers as to what should be mutable and what shouldn't. 

The input is items (features) which we want to store the original version, but then also a separate output version of that users would modify (they would never modify the original items, it is for reference and adding source links and such).   I'm also wondering if it makes sense to copy the input items by default into the output structure.

The process definition on the other hand should generally not be changed by the user (although I've hit a couple use cases where I modify the parameters for future tasks).

So basically the reason for splitting them up is because the user interacts with the Items and the Process definition in different ways. Although it probably makes sense to treat them in a similar way.  Make a deepcopy of the original paylod then a copy that can be changed by the user.